nebula3/gclient/net/SessionPool.py

```diff
@@ -0,0 +1,486 @@
+                status = 'BAD'
+            msg_list.append('[services: {}, status: {}]'.format(addr, status))
+        return ', '.join(msg_list)
+
+    def update_servers_status(self):
+        """update the servers' status"""
+        for address in self._addresses:
+            if self.ping(address):
+                self._addresses_status[address] = self.S_OK
+            else:
+                self._addresses_status[address] = self.S_BAD
+
+    def ping_sessions(self):
+        """ping all sessions in the pool"""
+        with self._lock:
+            for session in self._idle_sessions:
+                session.execute(r'RETURN "SESSION PING"')
+
+    def _get_idle_session(self):
+        """get a valid session from the pool idle list.
+
+        :return: Session
+        """
+        with self._lock:
+            if len(self._idle_sessions) > 0:
+                return self._idle_sessions.pop(0)
+            elif len(self._active_sessions) < self._configs.max_size:
+                return self._new_session()
+            else:
+                raise NoValidSessionException(
+                    'The total number of sessions reaches the pool max size {}'.format(
+                        self._configs.max_size
+                    )
+                )
+
+    def _new_session(self):
+        """construct a new session with the username and password in the pool.
+            also, the session is bound to the space specified in the configs.
+
+        :return: Session
+        """
+        if self._ssl_configs is not None:
+            raise RuntimeError('SSL is not supported yet')
+
+        self._pos = (self._pos + 1) % len(self._addresses)
+        next_addr_index = self._pos
+
+        # try to connect with a valid service address, the worst case it to iterate all addresses
+        retries = len(self._addresses)
+
+        while retries > 0:
+            addr = self._addresses[next_addr_index]
+            if self._addresses_status[addr] == self.S_OK:
+                connection = Connection()
+                try:
+                    connection.open(addr[0], addr[1], self._configs.timeout)
+                    auth_result = connection.authenticate(
+                        self._username, self._password
+                    )
+                    session = Session(connection, auth_result, self, False)
+
+                    # switch to the space specified in the configs
+                    resp = session.execute('USE {}'.format(self._space_name))
+                    if not resp.is_succeeded():
+                        raise RuntimeError(
+                            'Failed to get session, cannot set the session space to {} error: {} {}'.format(
+                                self._space_name, resp.error_code(), resp.error_msg()
+                            )
+                        )
+                    return session
+                except AuthFailedException as e:
+                    # if auth failed, close the pool
```

Comment:
What if somehow the authentication failed not due to bad credentials, but because it was something wrong with the metaD? Will closing the pool still be valid in this case?

Reply:
Nice catch, I'll add check for the error code to handle the credential case specifically. For non-credential errors, the pool should remain open.